lib/src/phoenix_presence.dart

```diff
@@ -0,0 +1,175 @@
+import 'dart:convert';
+import 'package:phoenix_wings/src/phoenix_channel.dart';
+
+class PhoenixPresence {
+  PhoenixChannel channel;
+  Map<String, dynamic> opts;
+  PresenceEvents events;
+  Map<String, Map<String, dynamic>> state;
+  List pendingDiffs = [];
+  String joinRef;
+  _PresenceCallers caller;
+
+  static void _noop(key, currentPresence, newPresence) {
+  }
+
+  PhoenixPresence(this.channel, {this.opts}) {
+    opts ??= {};
+    events = opts.containsKey('events') ? opts['events'] : PresenceEvents(PhoenixPresenceEvents.presenceState, PhoenixPresenceEvents.presenceDiff);
+    state = {};
+    joinRef = null;
+    caller = _PresenceCallers(_noop, _noop, () {});
+
+    this.channel.on(events.state, (newState, _ref, _joinRef) {
+      this.joinRef = this.channel.joinRef;
+      this.state = Map<String, Map<String, dynamic>>.from(syncState(this.state, newState, caller.onJoin, caller.onLeave));
+
+      this.pendingDiffs.forEach((diff) {
+        this.state = Map<String, Map<String, dynamic>>.from(syncDiff(this.state, diff, caller.onJoin, caller.onLeave));
+      });
+      this.pendingDiffs = [];
+      caller.onSync();
+    });
+
+    this.channel.on(events.diff, (diff, _ref, _joinRef) {
+      if (this.inPendingSyncState()) {
+        this.pendingDiffs.add(diff);
+      } else {
+        this.state = Map<String, Map<String, dynamic>>.from(syncDiff(this.state, diff, caller.onJoin, caller.onLeave));
+        caller.onSync();
+      }
+    });
+  }
+
+  onJoin(Function(dynamic key, dynamic currentPresence, dynamic newPresence) callback) => this.caller.onJoin = callback;
+
+  onLeave(Function(dynamic key, dynamic currentPresence, dynamic newPresence) callback) => this.caller.onLeave = callback;
+
+  onSync(Function() callback) => this.caller.onSync = callback;
+
+  list({Function by = null}) {
+    return _list(this.state, by);
+  }
+
+  inPendingSyncState(){
+    return this.joinRef == null || (this.joinRef != this.channel.joinRef);
+  }
+
+  static syncState(currentState, newState, onJoinCallback, onLeaveCallback){
+    var state = clone(currentState);
+    var joins = {};
+    var leaves = {};
+
+    state.forEach((key, presence) {
+      if (newState[key] == null) {
+        leaves[key] = presence;
+      }
+    });
+
+    newState.forEach((key, newPresence) {
+      var currentPresence = state[key];
```

Comment:
Consider early return rather than if/else

```dart
      var currentPresence = state[key];
      if (currentPresence == null) {
        return joins[key] = newPresence;
      }
      var newRefs = List<Map<String, dynamic>>.from(newPresence['metas'])
        .where((meta) => meta.containsKey('phx_ref'))
        .map((meta) => meta['phx_ref']);

      var curRefs = List<Map<String, dynamic>>.from(currentPresence['metas'])
        .where((meta) => meta.containsKey('phx_ref'))
        .map((meta) => meta['phx_ref']);

      var joinedMetas = newPresence['metas'].where((meta) => !curRefs.contains(meta['phx_ref']));
      var leftMetas = currentPresence['metas'].where((meta) => !newRefs.contains(meta['phx_ref']));

      if (joinedMetas.length > 0) {
        joins[key] = clone(newPresence);
        joins[key]['metas'] = List<Map<String, dynamic>>.from(joinedMetas);
      }

      if (leftMetas.length > 0) {
        leaves[key] = clone(currentPresence);
        leaves[key]['metas'] = List<Map<String, dynamic>>.from(leftMetas);
      }
```